tqdm ProgressBar implementation

cbouy opened the pull request:
Fixes #928 
Fixes #2504 

Changes made in this Pull Request:
 - Added the `ProgressBar` class which inherits the `tqdm.auto.tqdm` object. For now, the `disable` keyword takes precedence over `verbose`, i.e. if both are set to True the progress bar won't show. Setting `disable=True` will disable the progress bar as expected. Setting `verbose=False` will disable the progress bar.
- Deprecate `ProgressMeter` through warning and text
- Replace `ProgressMeter` with `ProgressBar`

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

cbouy commented:
Now it should be able to automatically detect Jupyter (notebook or lab) versus the rest and use the appropriate `tqdm` class.
For JupyterLab, it requires a few things beforehand:
```
conda install -c conda-forge nodejs
jupyter labextension install @jupyter-widgets/jupyterlab-manager
```

orbeckst commented on .appveyor.yml:
```diff
@@ -10,8 +10,8 @@ cache:
 environment:
     global:
         CONDA_CHANNELS: conda-forge
-        CONDA_DEPENDENCIES: pip setuptools wheel cython mock six biopython networkx joblib matplotlib scipy vs2015_runtime pytest mmtf-python GridDataFormats hypothesis pytest-cov codecov chemfiles
```
Why was chemfiles deleted?

cbouy replied:
That's an error from my part, my bad

orbeckst commented on .appveyor.yml:
```diff
@@ -10,8 +10,8 @@ cache:
-        CONDA_DEPENDENCIES: pip setuptools wheel cython mock six biopython networkx joblib matplotlib scipy vs2015_runtime pytest mmtf-python GridDataFormats hypothesis pytest-cov codecov chemfiles
-        PIP_DEPENDENCIES: gsd==1.9.3 duecredit parmed
+        CONDA_DEPENDENCIES: pip setuptools wheel cython mock six biopython networkx joblib matplotlib scipy vs2015_runtime pytest mmtf-python GridDataFormats hypothesis pytest-cov codecov
+        PIP_DEPENDENCIES: gsd==1.9.3 duecredit parmed tqdm
```
Is there no tqdm conda package for Windows? (Generally we prefer the conda package.)

cbouy replied:
Yes there is, good to know. I'll change that

orbeckst commented on package/MDAnalysis/lib/log.py:
```diff
@@ -394,3 +396,34 @@ def echo(self, step, **kwargs):
+
+# detect if running from jupyter, ipython or python
+try:
+    shell = get_ipython().__class__.__name__
```
Where does `get_ipython()` come from? Does it only exist in jupyter/ipython?

cbouy replied:
It comes from IPython and is automatically imported when running an ipython shell or jupyter notebook/lab

orbeckst commented on package/MDAnalysis/lib/log.py:
```diff
@@ -394,3 +396,34 @@ def echo(self, step, **kwargs):
+# detect if running from jupyter, ipython or python
+try:
+    shell = get_ipython().__class__.__name__
+except: # non-interactive shell
```
Can you be more precise about the exception that is raised, e.g., `NameError`?

Some comment on how you do the detection would make it clearer.

cbouy replied:
I noticed that tqdm actually has a class for automatic detection of this so I'll be changing the code to use their solution instead

orbeckst commented on package/MDAnalysis/lib/log.py:
```diff
@@ -394,3 +396,34 @@ def echo(self, step, **kwargs):
+        from tqdm import tqdm as _tqdm_class
+
+class ProgressBar(_tqdm_class):
+    r"""tqdm-like progress bar. Replaces the :class:`ProgressMeter` class to
```
This is placeholder text and eventually needs to be replace by something more meaningful.

If you leave out the doc string, will it inherit the one from `_tqdm_class`?

cbouy replied:
Yes, no doc-string = use the doc-string from the parent class. Should I do that instead ?

orbeckst replied:
yes

orbeckst commented:
@jbarnoud I am adding you to this issue as the original driving force behind #928

richardjgowers commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -177,13 +170,16 @@ def run(self, start=None, stop=None, step=None, verbose=None):
-        for i, ts in enumerate(
-                self._trajectory[self.start:self.stop:self.step]):
+        verbose = getattr(self, '_verbose', False)
+        n_frames = len(range(*self._trajectory.check_slice_indices(
```
Rather than calculating the n_frames, I thought that a trajectory slice knew it's length?  Ie `len(u.trajectory[::2])` should work, so maybe tqdm can infer the length of the traj slice

cbouy replied:
That would be true if there wasn't an `enumerate` around the traj slice, and we need it here ! But I just realised I can remove the `n_frames` for hbond_analysis.py, remove the `enumerate` in density.py, helanal.py and remove both `n_frames` and `enumerate` in pca.py. Thanks !

jbarnoud replied:
You should still use len on the sliced trajectory to get n_frames. It would make the code more readable and avoid potential issues we had mixing slices and range.

cbouy replied:
Makes sense ! I also just realised I can put the ProgressBar inside enumerate instead of the other way around, which makes it even simpler.

orbeckst commented:
The Python 2.7 version of the test is failing https://travis-ci.com/github/MDAnalysis/mdanalysis/jobs/298190141, perhaps something with comparing utf-8 strings.

cbouy commented:
@jbarnoud should I make a proper test module for the ProgressBar and ProgressMeter then ? I find it counter-intuitive to have the test for the ProgressMeter alongside the tests for the RMSD

jbarnoud commented:
I expect a test that creates a ProgresdMeter and makes sure the deprecation warning is issued. It should probably go with the tests of the log module.The point is to make sure that users that use the progress meter on its own actually see the warning, and that displaying it does not cause an error. We had issues like that in the past, where raising an exception was itself causing an error, because of format errors in the message. On 15 Mar 2020 17:44, Cédric Bouysset <notifications@github.com> wrote:
@jbarnoud should I make a proper test module for the ProgressBar and ProgressMeter then ? I find it counter-intuitive to have the test for the ProgressMeter alongside the tests for the RMSD

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

cbouy commented:
I don't understand the report from codecov, could someone explain what's missing ? I've added a new class, a test for this new class, I moved a test from test_rms.py to test_log.py and created a new test for the deprecation. There should be an increase in coverage or am I getting this wrong ?

orbeckst commented:
Codecov can be mysterious... 

Sorry for the long silence. Will have a look now.

orbeckst commented on package/MDAnalysis/analysis/hbonds/hbond_autocorrel.py:
```diff
@@ -392,12 +392,10 @@ def run(self, force=False):
-
-        for i, (start, stop) in enumerate(zip(self._starts, self._stops)):
-            pm.echo(i)
+        n_frames = len(self.u.trajectory)
```
`n_frames` is the wrong total, should be `self.nruns`

cbouy replied:
Oh yes ! my bad

orbeckst commented on package/MDAnalysis/analysis/hbonds/hbond_autocorrel.py:
```diff
@@ -392,12 +392,10 @@ def run(self, force=False):
-            pm.echo(i)
+        n_frames = len(self.u.trajectory)
+        for i, (start, stop) in ProgressBar(enumerate(zip(self._starts,
+                                            self._stops)), total=n_frames,
```
```suggestion
                                            self._stops)), total=self.nruns,
```

orbeckst commented:
> Codecov can be mysterious...

The above was a bit flippant but what I wanted to say is that it's not always clear that codecov is showing us what we expect to see and although we like to see everything green in the CI, we do overrule Codecov once we are happy.

jbarnoud commented:
@lilyminium noticed in https://github.com/MDAnalysis/mdanalysis/pull/2631#issuecomment-601649236 that the API changes. It probably should not.

lilyminium commented:
https://github.com/MDAnalysis/mdanalysis/pull/2631#issuecomment-601649236 was phrased poorly -- as far as I can tell this does not change the API, except that `AnalysisBase._pm` no longer exists. #2631 did alter the arguments given to `_setup_frames`. This PR also fixes #2504.

orbeckst commented on package/CHANGELOG:
```diff
@@ -17,7 +17,7 @@ mm/dd/yy richardjgowers, kain88-de, lilyminium, p-j-smith, bdice, joaomcteixeira
          Yuan-Yu, xiki-tempula, HTian1997, Iv-Hristov, hmacdope, AnshulAngaria,
          ss62171, Luthaf, yuxuanzhuang, abhishandy, mlnance, shfrz, orbeckst,
-         wvandertoorn
+         wvandertoorn, cbouy
```
Could you please update the CHANGELOG to also say that #2504 is fixed?

lilyminium replied:
This wasn’t in the scope of the original PR, but if you do, could you add a test that the verbose keyword does actually toggle the ProgressBar?

cbouy replied:
Done !

orbeckst commented:
@lilyminium @jbarnoud  thanks for pointing out that this PR fixes #2504, too https://github.com/MDAnalysis/mdanalysis/pull/2631#issuecomment-601649236.

cbouy commented:
Thank you @orbeckst @jbarnoud @richardjgowers for your help ! And thanks @lilyminium for mentioning the other issue !

orbeckst commented:
One of the macOS jobs failed right at the start. I restarted https://travis-ci.com/github/MDAnalysis/mdanalysis/jobs/300680256 – hopefully this will all look good in an hour.

orbeckst commented:
@cbouy very nice contribution!
